pkg/mls/api/v1/service.go

```diff
@@ -345,8 +345,32 @@ func (s *Service) SubscribeGroupMessages(req *mlsv1.SubscribeGroupMessagesReques
 	// See: https://github.com/xmtp/libxmtp/pull/58
 	_ = stream.SendHeader(metadata.Pairs("subscribed", "true"))
 
-	var streamLock sync.Mutex
+	streamed := map[string]*mlsv1.GroupMessage{}
+	var streamingLock sync.Mutex
+	streamMessages := func(msgs []*mlsv1.GroupMessage) {
+		streamingLock.Lock()
+		defer streamingLock.Unlock()
+
+		for _, msg := range msgs {
+			if msg.GetV1() == nil {
+				continue
+			}
+			encodedId := fmt.Sprintf("%x", msg.GetV1().Id)
+			if _, ok := streamed[encodedId]; ok {
+				log.Debug("skipping already streamed message", zap.String("id", encodedId))
+				continue
+			}
+			err := stream.Send(msg)
+			if err != nil {
+				log.Error("error streaming group message", zap.Error(err))
+			}
+			streamed[encodedId] = msg
+		}
+	}
+
 	for _, filter := range req.Filters {
+		filter := filter
```

Comment:
Why is this assignment necessary?

Reply:
This [is needed](https://eli.thegreenplace.net/2019/go-internals-capturing-loop-variables-in-closures/) because `filter` is used within the goroutine function below. Alternatively you can have the function accept it as an arg. This is a pretty annoying footgun in go, but fortunately linters complain about it these days if missing with
> loopclosure: loop variable filter captured by func literal (govet)

Reply:
Go has some very unintuitive semantics around loops and closures. Without this, you'd end up always referencing the last fixture in the list. 

They are finally [fixing it](https://go.dev/wiki/LoopvarExperiment) but it's not widely enabled.